src/nn/wasm/PreparedModel.js

```diff
@@ -111,24 +113,148 @@ export default class PreparedModel {
+        throw new Error("Unsupported type of tensor for fused activation function.");
+      }
+    }
+
+    function GetQuantizedConvolutionMultipler(input_scale, filter_scale, 
+                                              bias_scale, output_scale) {
+      let input_product_scale = input_scale * filter_scale;
+      // const float bias_scale = biasShape.scale;
+      // const float output_scale = outputShape.scale;
+
+      // The following conditions must be guaranteed by the training pipeline.
+      OPS_CHECK(Math.abs(input_product_scale - bias_scale) <=
+                (1e-6 * Math.min(input_product_scale, bias_scale)));
+      OPS_CHECK(input_product_scale >= 0);
+      OPS_CHECK(input_product_scale < output_scale);
+      let multiplier = input_product_scale / output_scale;
+      return multiplier;
+    }
+
+    function QuantizeMultiplierSmallerThanOne(double_multiplier) {
+      let quantized_multiplier, right_shift, q;
+      OPS_CHECK(double_multiplier >= 0.);
+      OPS_CHECK(double_multiplier < 1.);
+      if (double_multiplier === 0.) {
+        quantized_multiplier = 0;
+        right_shift = 0;
+        return [quantized_multiplier, right_shift];
+      }
+      OPS_CHECK(double_multiplier > 0.);
+      [q, right_shift] = frexp(double_multiplier);
+      right_shift *= -1;
+      quantized_multiplier = Math.round(q * -(1 << 31));
+      OPS_CHECK(quantized_multiplier <= -(1 << 31));
+      if (quantized_multiplier == -(1 << 31)) {
+        quantized_multiplier /= 2;
+        --right_shift;
+      }
+      OPS_CHECK(right_shift >= 0);
+      OPS_CHECK(quantized_multiplier <= nn_ops.INT32_MAX);
+      // quantized_multiplier = static_cast<int32_t>(q_fixed);
+      return [quantized_multiplier, right_shift];
+    }
+
+    function QuantizeMultiplierGreaterThanOne(double_multiplier) {
+      let quantized_multiplier, left_shift, q;
+      OPS_CHECK(double_multiplier > 1.);
+      [q, left_shift] = frexp(double_multiplier);
+      quantized_multiplier = Math.round(q * -(1 << 31));
+      OPS_CHECK(quantized_multiplier <= -(1 << 31));
+      if (quantized_multiplier == -(1 << 31)) {
+        quantized_multiplier /= 2;
+        ++left_shift;
+      }
+      OPS_CHECK(left_shift >= 0);
+      OPS_CHECK(quantized_multiplier <= nn_ops.INT32_MAX);
+      return [quantized_multiplier, left_shift];
+    }
+
+    function CalculateInputRadius(input_integer_bits, input_left_shift) {
+      let max_input_rescaled = 1.0 * ((1 << input_integer_bits) - 1) *
+                                        (1 << (31 - input_integer_bits)) /
+                                        (1 << input_left_shift);
+      // Tighten bound using floor.  Suppose that we could use the exact value.
+      // After scaling the difference, the result would be at the maximum.  Thus we
+      // must ensure that our value has lower magnitude.
+      return Math.floor(max_input_rescaled);
+    }
+
+    function frexp (arg) {
```

Comment:
Is the code converted from C counterpart? If yes, please add reference to C function. Thanks!